sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,707 @@
+	s.Logger.Info("sharding bucket", "bucket_key", bucket.Key, "num_shards", numShards)
+	defer s.Logger.Info("sharding bucket process exited", "bucket_key", bucket.Key)
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
+		lock := locksutil.LockForKey(s.storageLocks, cacheKey+shardKey)
+		shardedBucket := &LockedBucket{
+			LockEntry: lock,
+			Bucket: &Bucket{
+				Key:       fmt.Sprintf("%s/%s", bucket.Key, shardKey),
+				ItemMap:   make(map[string][]byte),
+				HasShards: false,
+			},
+		}
+		shards[shardKey] = shardedBucket
+		s.Logger.Trace("created shard", "shard_key", shardKey)
+	}
+
+	s.Logger.Debug("resilvering items")
+
+	parentPrefix := s.GetCacheKey(bucket.Key)
+	// Resilver the items
+	for id, v := range bucket.ItemMap {
+		key := GetItemIDHash(id)
+		shardKey := strings.TrimPrefix(key, parentPrefix)[0 : s.BucketShardBits/4]
+		s.Logger.Trace("resilvering item", "parent_prefix", parentPrefix, "item_id", id, "shard", shardKey)
+		// Sanity check
+		childBucket, ok := shards[shardKey]
+		if !ok {
+			// We didn't complete sharding so don't make other parts of the
+			// code think that it completed
+			s.Logger.Error("failed to find sharded storagepacker bucket", "bucket_key", bucket.Key, "shard", shardKey)
+			return errors.New("failed to shard storagepacker bucket")
+		}
+		childBucket.ItemMap[id] = v
+	}
+
+	s.Logger.Debug("storing sharded buckets")
+
+	// Ensure we can write all of these buckets. Create a cleanup function if not.
+	retErr := new(multierror.Error)
+	cleanupStorage := func() {
+		for _, v := range shards {
+			if err := s.BucketStorageView.Delete(ctx, v.Key); err != nil {
+				retErr = multierror.Append(retErr, err)
+				// Don't exit out, clean up as much as possible
+			}
+		}
+	}
+	for k, v := range shards {
+		s.Logger.Trace("storing bucket", "shard", k)
+		// Avoid queuing mode, go directly to "store" not "persist"
+		if err := s.storeBucket(ctx, v, 1); err != nil {
+			s.Logger.Debug("encountered error", "shard", k)
+			retErr = multierror.Append(retErr, err)
+			cleanupStorage()
+			return retErr
+		}
+	}
+
+	// Update the original and persist it.
+	// Prior to this point, version N is in storage and we revert back to it.
+	// After this point, storage is in version N+1 but the cache has not yet
+	// been updated; anybody trying to access one of the keys will hit
+	// this bucket and block on the lock.
+	origBucketItemMap := bucket.ItemMap
+	bucket.ItemMap = nil
+	bucket.HasShards = true
+	if err := s.storeBucket(ctx, bucket, 1); err != nil {
+		retErr = multierror.Append(retErr, err)
+		bucket.ItemMap = origBucketItemMap
+		bucket.HasShards = false
+		cleanupStorage()
+		return retErr
+	}
+
+	// Add to the cache.
+	s.Logger.Debug("updating cache")
+	s.bucketsCacheLock.Lock()
+	for _, v := range shards {
+		s.bucketsCache.Insert(s.GetCacheKey(v.Key), v)
+	}
+	s.bucketsCacheLock.Unlock()
+
+	return retErr.ErrorOrNil()
```

Comment:
FWIW, I think you can just return `retErr` here and it will behave as you would expect.

Reply:
Hmm... I think that would be the case if I we hadn't done `retErr := new(multierror.Error)`, it will return an error object with zero errors.

https://play.golang.org/p/MNSY05oQGcF

It seems possible from the docs https://github.com/hashicorp/go-multierror to use multierror.Append if the first argument is an uninitialized error variable, though, so we could do it like this:

https://play.golang.org/p/k8shsd_6MFj

Reply:
Ah yes! I hadn't caught that the multierror was in the mix. :+1: